src/software/ai/passing/receiver_position_generator.hpp

```diff
@@ -0,0 +1,344 @@
+     * @return A vector of the top num_positions zones
+     */
+    std::vector<ZoneEnum> getTopZones(
+        const std::map<ZoneEnum, PassWithRating> &best_receiving_positions,
+        unsigned int num_positions, const Point &pass_origin,
+        const std::vector<Point> &existing_receiver_positions);
+
+    /**
+     * Helper method for visualizing the best receiving positions and zones
+     * @param best_receiving_positions The best receiving positions and their ratings
+     * @param top_zones The ranked top zones to visualize
+     */
+    void visualizeBestReceivingPositionsAndZones(
+        const std::map<ZoneEnum, PassWithRating> &best_receiving_positions,
+        const std::vector<ZoneEnum> &top_zones);
+
+    // Pitch division
+    std::shared_ptr<const FieldPitchDivision<ZoneEnum>> pitch_division_;
+
+    // The best receiving position for each zone from the previous iteration
+    std::map<ZoneEnum, Point> prev_best_receiving_positions;
+
+    // Passing configuration
+    TbotsProto::PassingConfig passing_config_;
+
+    // A vector of shapes that will be visualized
+    std::vector<TbotsProto::DebugShapes::DebugShape> debug_shapes;
+
+    // A random number generator for use across the class
+    std::mt19937 random_num_gen_;
+
+    // The random seed to initialize the random number generator
+    static constexpr int RNG_SEED = 1010;
+};
+
+template <class ZoneEnum>
+ReceiverPositionGenerator<ZoneEnum>::ReceiverPositionGenerator(
+    std::shared_ptr<const FieldPitchDivision<ZoneEnum>> pitch_division,
+    TbotsProto::PassingConfig passing_config)
+    : pitch_division_(pitch_division),
+      passing_config_(passing_config),
+      random_num_gen_(RNG_SEED)
+{
+}
+
+template <class ZoneEnum>
+std::vector<Point> ReceiverPositionGenerator<ZoneEnum>::getBestReceivingPositions(
+    const World &world, unsigned int num_positions,
+    const std::vector<Point> &existing_receiver_positions,
+    const std::optional<Point> &pass_origin_override)
+{
+    std::map<ZoneEnum, PassWithRating> best_receiving_positions;
+    debug_shapes.clear();
+
+    Point pass_origin           = pass_origin_override.value_or(world.ball().position());
+    const auto &receiver_config = passing_config_.receiver_position_generator_config();
+
+    // Verify that the number of receiver positions requested is valid
+    if (num_positions >
+        world.friendlyTeam().numRobots() - existing_receiver_positions.size())
+    {
+        LOG(WARNING) << "Not enough friendly robots to assign " << num_positions
+                     << " receiver positions. Assigning "
+                     << world.friendlyTeam().numRobots() -
+                            existing_receiver_positions.size()
+                     << " receiver positions instead";
+        num_positions = static_cast<unsigned int>(world.friendlyTeam().numRobots() -
+                                                  existing_receiver_positions.size());
```

Comment:
does this have strange implications?

what happens if we are playing the game and one of our robots foul out? Will this cause this to overflow?

Reply:
Are you implying a scenario where `existing_receiver_positions.size() > world.friendlyTeam().numRobots()`? I feel like the caller is the one at fault at that point. The caller should not just add a bunch of "existing receivers" without confirming that there can actually be that many robots on the field.